docs/Operators.md

```diff
@@ -7611,17 +7611,213 @@ Other versions of this operator: <a href="Changelog.md#GreaterOrEqual-12">12</a>
-  See also in [torch.nn.functional.grid_sample](https://pytorch.org/docs/master/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample).
+**Note: This is intentionally a direct edit to Operators.md even though it is
+auto-generated which is more convenient during editing. The real patch
+will do this the proper way. Also, this is currently a draft
+supporting the issue reported on GridSample. The draft simply assumes
+certain decisions that are up for discussion in the filed issue.**
+
+GridSample is conceptually similar to gather_nd but allows access to data at non-integer
+coordinates by interpolating from nearby integer coordinates. GridSample also
+allows access at coordinates beyond the size of the input.
+
+Given a 4D `input` tensor with dimensions `[N,C,H,W] ` and a 4D `grid` tensor
+with dimensions `[N,W',H',2]`, the 4D output tensor of `GridSample` has
+dimensions `[N,C,W',H']`. Conceptually, the output element at `[n,c,row',col']`
+equals `input[n,c,row,col]` where `(row, col)` is calculated from the 2-element
+vector `(x,y)` that is equal to `grid(n,row',col')`. Conceptually, that is what
+is going on. The details are more involved as described below.
+
+See also in [torch.nn.functional.grid_sample](https://pytorch.org/docs/master/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample). Currently, ONNX GridSample requires 4D tensors.
+
+#### Denormalization of coordinates depending on `align_corners`
+
+A coordinate vector is in-bounds if it stays within the dimension sizes of
+`input` in the height and width dimensions. A coordinate vector in GridSample is
+not required to be in-bounds. This section concerns in-bounds coordinate vectors
+while out-of-bounds coordinate vectors are handled in a later section.
+
+The in-bounds coordinate vectors in `grid` are normalized to the range `[-1,1]`.
+The indices of `input` are in a range `[0,W)` and `[0,H)`, respectively, so we
+need to denormalize the coordinates in `grid` to put them into a range more
+appropriate for indexing into `input`. The calculation to do the denormalization
+depends on the value of `align_corners`.
+
+If `align_corners` is `0` (i.e. `false`), then for the horizontal (left-right)
+dimension, `-1` refers to the left border of the left-most column of pixels while
+`1` refers to the right border of the right-most column of pixels. The formula for
+denormalization of a coordinate `x` is then:
+
+  `denormalize(x, W, false) = ((x + 1) * W - 1) / 2` .
+
+If `align_corners` is `1` (i.e. `true`), then for the horizontal (left-right)
+dimension, `-1` refers to the center of the left-most column of pixels while `1`
+refers to the center of the right-most column of pixels. The formula for
+denormalization of a coordinate `x` is then:
+
+  `denormalize(x, W, true) = (x + 1) * (W - 1) / 2` .
+
+The normalized coordinates in `grid` are notated as `(x, y)` while the
```

Comment:
Relation between x, y coordinate and row, col in data matrix are understandable in general. I suggest to either use x, y[, z] in op description. Avoiding using row, col shall reduce document length.

Reply:
I think the description is good in its current form. x and y refer to normalized-coordinates in the range [-1, 1], IIUC, and the denormalize function computes the row, col used to index into the input matrix. So, they are distinct.

Reply:
Yes, it's unfortunate that one cannot avoid using both concepts since the op inherently involves both.